packages/components/src/draggable/index.native.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * External dependencies
+ */
+import { Gesture, GestureDetector } from 'react-native-gesture-handler';
+import Animated, { useSharedValue } from 'react-native-reanimated';
+
+/**
+ * Draggable component
+ *
+ * @param {Object}                                      props                       Component props.
+ * @param {JSX.Element}                                 props.children              Children to be rendered.
+ * @param {?number}                                     props.maxDistance           Maximum distance, that defines how far the finger is allowed to travel during a long press gesture.
+ * @param {?number}                                     props.minDuration           Minimum time, that a finger must remain pressed on the corresponding view.
+ * @param {Function}                                    props.onDragEnd             Callback when dragging happens over an element.
+ * @param {Function}                                    props.onDragOver            Callback when dragging ends.
+ * @param {Function}                                    props.onDragStart           Callback when dragging stars.
+ * @param {import('react-native-reanimated').StyleProp} props.wrapperAnimatedStyles Animated styles for the wrapper component.
+ *
+ * @return {JSX.Element} The component to be rendered.
+ */
+export default function Draggable( {
+	children,
+	maxDistance = 1000,
+	minDuration = 500,
+	onDragEnd,
+	onDragOver,
+	onDragStart,
+	wrapperAnimatedStyles,
+} ) {
+	const isDragging = useSharedValue( false );
+	const hasPanStarted = useSharedValue( false );
+
+	const longPressGesture = Gesture.LongPress()
+		.onStart( ( ev ) => {
+			'worklet';
+			isDragging.value = true;
+
+			if ( onDragStart ) {
+				onDragStart( ev );
+			}
+		} )
+		.onEnd( () => {
+			if ( ! hasPanStarted.value && isDragging.value ) {
+				isDragging.value = false;
+
+				if ( onDragEnd ) {
+					onDragEnd();
+				}
+			}
+		} )
+		.maxDistance( maxDistance )
+		.minDuration( minDuration )
+		.shouldCancelWhenOutside( false );
+
+	const panGesture = Gesture.Pan()
+		.manualActivation( true )
+		.onTouchesMove( ( _, state ) => {
+			'worklet';
+			if ( isDragging.value ) {
+				hasPanStarted.value = true;
+				state.activate();
+			} else {
+				state.fail();
```

Comment:
After further testing these changes and `BlockDraggable` component, I noticed that on Android the callback `onTouchesMove` can be called before `LongPress.onStart`. Besides, setting the state of the pan gesture to `failed` prevents this callback to be triggered again until releasing the touch. All these lead to a situation where the draggable gesture doesn't work if the previous is fulfilled, which as far as I tested can be easily reproduced.

Being that said, I think we might consider dropping setting the state to `failed` to prevent the issue described above.

Reply:
Btw, I managed to fix this by applying the following patch:
```diff
diff --git a/packages/components/src/draggable/index.native.js b/packages/components/src/draggable/index.native.js
index a9543a0d8fd87ea38c4ff07fce66b67bf103ad2b..85037fe80d71e4503feb737195ab2b70f753d10c 100644
--- a/packages/components/src/draggable/index.native.js
+++ b/packages/components/src/draggable/index.native.js
@@ -59,8 +59,6 @@ export default function Draggable( {
 			if ( isDragging.value ) {
 				hasPanStarted.value = true;
 				state.activate();
-			} else {
-				state.fail();
 			}
 		} )
 		.onUpdate( ( ev ) => {
@@ -80,9 +78,10 @@ export default function Draggable( {
 			isDragging.value = false;
 			hasPanStarted.value = false;
 		} )
+		.shouldCancelWhenOutside( false )
 		.simultaneousWithExternalGesture( longPressGesture );
 
-	const dragHandler = Gesture.Race( panGesture, longPressGesture );
+	const dragHandler = Gesture.Simultaneous( panGesture, longPressGesture );
 
 	return (
 		<GestureDetector gesture={ dragHandler }>

```

Reply:
Nice findings! I will apply the mentioned changes 👍

Reply:
I'm seeing that removing `state.fail();` causes the scroll on iOS to not work 😅 . Can you reproduce this? On Android the scroll keeps working.

Maybe we could only add `state.fail()` for iOS 🤔

Reply:
> I'm seeing that removing `state.fail();` causes the scroll on iOS to not work 😅 . Can you reproduce this? On Android the scroll keeps working.
> 
> Maybe we could only add `state.fail()` for iOS 🤔

Right, I thought I tested on iOS but I've just reproduced the same you mentioned. At this point, I think the better would be to check the platform and call `state.fail()` only for the iOS case.

Reply:
Cool, I'll add the iOS check 👍